logging/google/cloud/logging/handlers/transports/background_thread.py

```diff
@@ -19,137 +19,224 @@
+def _get_many(q, max_items=None):
+    """Get multiple items from a Queue.
+
+    Gets at least one (blocking) and at most :param:`max_items` items
+    (non-blocking) from a given Queue. Does not mark the items as done.
 
-    Writes entries to the logger API.
+    :type q: ~queue.Queue
+    :param q: The Queue to get items from.
 
-    This class reuses a single :class:`Batch` method to write successive
-    entries.
+    :type max_items: int
+    :param max_items: The maximum number of items to get. If ``None``, then all
+        available items in the queue are returned.
 
-    Currently, the only public methods are constructing it (which also starts
-    it) and enqueuing :class:`Logger` (record, message) pairs.
+    :rtype: Sequence
+    :returns: A sequence of items retrieved from the queue.
     """
+    # Always return at least one item.
+    items = [q.get()]
+    while max_items is None or len(items) < max_items:
+        try:
+            items.append(q.get_nowait())
+        except queue.Empty:
+            break
+    return items
 
-    def __init__(self, logger):
-        self.started = False
-        self.stopping = False
-        self.stopped = False
 
-        # _entries_condition is used to signal from the main thread whether
-        # there are any waiting queued logger entries to be written
-        self._entries_condition = threading.Condition()
+class _Worker(object):
+    """A background thread that writes batches of log entries."""
 
-        # _stop_condition is used to signal from the worker thread to the
-        # main thread that it's finished its last entries
-        self._stop_condition = threading.Condition()
+    def __init__(self, cloud_logger, grace_period=_DEFAULT_GRACE_PERIOD,
+                 max_batch_size=_DEFAULT_MAX_BATCH_SIZE):
+        """
+        The background thread is started automatically.
 
-        # This object continually reuses the same :class:`Batch` object to
-        # write multiple entries at the same time.
-        self.logger = logger
-        self.batch = self.logger.batch()
+        :type cloud_logger: ~google.cloud.logging.logger.Logger
+        :param cloud_logger: The logger to send entries to.
 
-        self._thread = None
+        :type grace_period: float
+        :param grace_period: The amount of time to wait for pending logs to
+            be submitted when the process is shutting down.
 
-        # Number in seconds of  how long to wait for worker to send remaining
-        self._stop_timeout = 5
+        :type max_batch_size: int
+        :param max_batch_size: The maximum number of items to send at a time
+            in the background thread.
+        """
+        self._cloud_logger = cloud_logger
+        self._grace_period = grace_period
+        self._max_batch_size = max_batch_size
+        self._queue = queue.Queue(-1)
+        self._operational_lock = threading.Lock()
+        self._thread = None
+        self.start()
```

Comment:
:-1: on this side effect in the constructor. Why can't the `__init__` caller also `start`? Or invoke a factory constructor that also starts?